internal/handler/control.go

```diff
@@ -1,22 +1,146 @@
 // -*- Mode: Go; indent-tabs-mode: t -*-
 //
 // Copyright (C) 2017-2018 Canonical Ltd
+// Copyright (C) 2020 IOTech Ltd
 //
 // SPDX-License-Identifier: Apache-2.0
 
 package handler
 
 import (
+	"context"
 	"fmt"
+	"io"
+	"net/http"
+	"regexp"
+	"sync/atomic"
+	"time"
 
+	"github.com/edgexfoundry/device-sdk-go/internal/cache"
 	"github.com/edgexfoundry/device-sdk-go/internal/common"
+	"github.com/edgexfoundry/device-sdk-go/pkg/models"
+	contract "github.com/edgexfoundry/go-mod-core-contracts/models"
+	"github.com/google/uuid"
 )
 
-func DiscoveryHandler(requestMap map[string]string) {
-	common.LoggingClient.Info(fmt.Sprintf("service: discovery request"))
-}
+var (
+	id     string
+	locker uint32
+)
 
 func TransformHandler(requestMap map[string]string) (map[string]string, common.AppError) {
 	common.LoggingClient.Info(fmt.Sprintf("service: transform request: transformData: %s", requestMap["transformData"]))
 	return requestMap, nil
 }
+
+func DiscoveryHandler(w http.ResponseWriter) {
+	if id == "" {
+		id = uuid.New().String()
+	}
+
+	if w != nil {
+		msg := fmt.Sprintf("Discovery triggered or already running, id = %s", id)
+		w.WriteHeader(http.StatusAccepted) //status=202
+		_, _ = io.WriteString(w, msg)
+	}
+
+	if !atomic.CompareAndSwapUint32(&locker, 0, 1) {
```

Comment:
This probably wouldn't be my first choice to handle ensuring this is a singleton operation. Couldn't you just use a standard mutex for this? If you want to keep it, please at least add a comment explaining this code.

Reply:
I still think you should change this. The godoc for the atomic package says:

> These functions require great care to be used correctly. Except for special, low-level applications, synchronization is better done with channels or the facilities of the sync package. Share memory by communicating; don't communicate by sharing memory.

Here's an alternate [implementation](https://github.com/tonyespy/device-sdk-go/commit/35d5133afaba6edb247191f6d48b8e550c66cc49) using `sync.Mutex` that I pushed to a [branch](https://github.com/tonyespy/device-sdk-go/tree/awe-discover) in my repo. **Note** - as mentioned in my comments, you also need to guard the global id variable too, so maybe it makes sense putting in a struct along with the `busy` flag.